### PR TITLE
Disregard static abstract methods as entrypoint candidates

### LINKED ISSUE #66516: "abstract void Main()" compiles successfully
**Version Used**: .NET 7.0

**Steps to Reproduce**:
1. Create console application
2. Write interface with `static abstract void Main()`
3. `dotnet run`

https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgZgAIBLKYAUzADMBDAYwqIGEiBvPIro1TANiK0ARgGdgYBsB4AWIgFlaZABQBKANx4AvkA

**Expected Behavior**:
Compilation fails with some error.

**Actual Behavior**:
Program compiles successfully, and exception is thrown when running program.

### PR DESCRIPTION
**Pull request by john-h-k**
Fixes #66516

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/Emit/EntryPointTests.cs [hunk 1]
```diff
@@ -1642,5 +1642,19 @@ class Program
...
                 Diagnostic(ErrorCode.ERR_NoEntryPoint));
         }
+
+        [Fact]
+        public void AbstractEntryPoint()
+        {
+            var source = @"
+interface Program
+{
+    static abstract void Main();
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/Emit/EntryPointTests.cs:1652]: Let's also include a positive test when you implement the `Program` interface and run it (have the impl print 1 or something trivial). To do this, you'll want to call `CompileAndVerify`, and then use `ExecutionConditionUtil.IsCoreClr` to pass in either the expected output or null (to skip running the binary) based on whether the test is executing on a framework that can run static abstracts.

**Review by 333fred — COMMENTED** [review]
Overall looks good, just a couple of minor test comments (commit 1)

File: src/Compilers/CSharp/Test/Emit/Emit/EntryPointTests.cs [hunk 2]
```diff
@@ -1642,5 +1642,41 @@ class Program
...
+using System;
+
+interface IProgram
+{
+    static abstract void Main();
+}
+
+class Program : IProgram
+{
+    public static void Main() => Console.WriteLine(""Hello World!"");
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/Emit/EntryPointTests.cs:1674]: Sorry for the continual requests, but finally let's an an explicit implementation :).

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs [hunk 3]
```diff
@@ -712,7 +712,7 @@ internal bool IsEntryPointCandidate
                     return false;
                 }
 
-                return IsStatic && Name == WellKnownMemberNames.EntryPointMethodName;
+                return IsStatic && !IsAbstract && Name == WellKnownMemberNames.EntryPointMethodName;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs:715]: >!IsAbstract [](http://example.com/codeflow?start=35&length=11)

Is `virtual` fine?

File: src/Compilers/CSharp/Test/Emit/Emit/EntryPointTests.cs [hunk 4]
```diff
@@ -1642,5 +1642,41 @@ class Program
...
                 Diagnostic(ErrorCode.ERR_NoEntryPoint));
         }
+
+        [Fact]
+        public void AbstractEntryPointWithoutImplementation()
+        {
+            var source = @"
+interface IProgram
+{
+    static abstract void Main();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/EntryPointTests.cs:1652]: >abstract [](http://example.com/codeflow?start=11&length=8)

Please add a test with a virtual method.

File: src/Compilers/CSharp/Test/Emit/Emit/EntryPointTests.cs [hunk 5]
```diff
@@ -1642,5 +1642,41 @@ class Program
...
+}";
+
+            var compilation = CreateCompilation(source, options: TestOptions.DebugExe, targetFramework: TargetFramework.Net70);
+            compilation.VerifyDiagnostics(
+                // error CS5001: Program does not contain a static 'Main' method suitable for an entry point
+                Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1));
+        }
+
+        [Fact]
+        public void AbstractEntryPointWithImplementation()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/EntryPointTests.cs:1662]: >AbstractEntryPointWithImplementation [](http://example.com/codeflow?start=20&length=36)

It looks like this test is failing

File: src/Compilers/CSharp/Test/Emit/Emit/EntryPointTests.cs [hunk 6]
```diff
@@ -1642,5 +1642,41 @@ class Program
                 // error CS5001: Program does not contain a static 'Main' method suitable for an entry point
                 Diagnostic(ErrorCode.ERR_NoEntryPoint));
         }
+
+        [Fact]
+        public void AbstractEntryPointWithoutImplementation()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/EntryPointTests.cs:1647]: >AbstractEntryPointWithoutImplementation [](http://example.com/codeflow?start=20&length=39)

I suggest moving the tests to src\Compilers\CSharp\Test\Symbol\Symbols\StaticAbstractMembersInInterfacesTests.cs and following the pattern used in that file for target framework and expected output in tests.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 3)

**Comment by 333fred** [comment]
@john-h-k please don't force push to PRs that are in review, it eliminates the ability to diff review your responses to feedback.

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs [hunk 7]
```diff
@@ -33432,5 +33432,121 @@ class C23_2 : I2<C23_2>, I3<C23_2>
                     Diagnostic(ErrorCode.ERR_DuplicateConversionInClass, "int").WithArguments("C23_2").WithLocation(80, 45)
                 );
         }
+
+        [Fact]
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:33436]: Consider putting a `WorkItem` attribute with a link to the bug these are fixing.

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs [hunk 8]
```diff
@@ -33432,5 +33432,127 @@ class C23_2 : I2<C23_2>, I3<C23_2>
...
+                {
+                    static abstract void Main();
+                }
+
+                class Program : IProgram
+                {
+                    public static void Main() => Console.WriteLine("Hello World!");
+                }
+                """,
+                expectedOutput: ExecutionConditionUtil.IsCoreClr ? "Hello World!" : null,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:33485]: >ExecutionConditionUtil.IsCoreClr [](http://example.com/codeflow?start=32&length=32)

```Execute(false)```?

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs [hunk 9]
```diff
@@ -33432,5 +33432,127 @@ class C23_2 : I2<C23_2>, I3<C23_2>
...
+                {
+                    static virtual void Main() => Console.WriteLine("Hello World!, from IProgram");
+                }
+
+                class Program : IProgram
+                {
+                    public static void Main() => Console.WriteLine("Hello World!");
+                }
+                """,
+                expectedOutput: ExecutionConditionUtil.IsCoreClr ? "Hello World!" : null,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:33530]: >ExecutionConditionUtil.IsCoreClr [](http://example.com/codeflow?start=32&length=32)

```Execute(false)```?

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs [hunk 10]
```diff
@@ -33432,5 +33432,127 @@ class C23_2 : I2<C23_2>, I3<C23_2>
...
+                    // error CS5001: Program does not contain a static 'Main' method suitable for an entry point
+                    Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1)
+                );
+        }
+
+        [Fact]
+        [WorkItem(66516, "https://github.com/dotnet/roslyn/issues/66516")]
+        public void AbstractEntryPointWithImplementation()
+        {
+            CompileAndVerify("""
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:33472]: >CompileAndVerify [](http://example.com/codeflow?start=12&length=16)

It looks like we need to skip verification - ```verify: Verification.Skipped```

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs [hunk 11]
```diff
@@ -33432,5 +33432,127 @@ class C23_2 : I2<C23_2>, I3<C23_2>
...
+                    // error CS5001: Program does not contain a static 'Main' method suitable for an entry point
+                    Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1)
+                );
+        }
+
+        [Fact]
+        [WorkItem(66516, "https://github.com/dotnet/roslyn/issues/66516")]
+        public void VirtualEntryPointWithImplementation()
+        {
+            CompileAndVerify("""
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:33517]: >CompileAndVerify [](http://example.com/codeflow?start=12&length=16)

It looks like we need to skip verification - ```verify: Verification.Skipped```

**Comment by AlekseyTs** [comment]
Done with review pass (commit 5)

**Comment by jcouv** [comment]
@john-h-k  We've had some CI issues. Could you merge latest bits from the `main` branch into this PR? (I do something like `git fetch origin main` then `git merge origin/main` then push the commit)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 7)

**Comment by AlekseyTs** [comment]
@john-h-k Thanks for the contribution

**Pull request merged**